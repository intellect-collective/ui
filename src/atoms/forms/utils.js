/**
 * Normalize a set of items for use in select or auto-complete inputs.
 *
 * @param {Array|Object} items - An array of title-value objects, or an object
 *         to be broken down
 * @returns {Array} - An array of title-value objects
 */
export const normalizeItems = (items) => {
    // If we received an array of options, they should already be formatted
    // correctly
    if (Array.isArray(items)) {
        return items
            .map(({ title, value }) => ({
                title,
                value,
                key: title
            }));
    }
    return Object.keys(items)
        .map((opt) => {
            return {
                title: items[opt],
                value: opt,
                key: items[opt]
            };
        });
};
