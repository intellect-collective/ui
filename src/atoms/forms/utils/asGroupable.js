import React from 'react';
import PropTypes from 'prop-types';
import { FieldGroupContext } from '../FieldGroup';

/**
 * The `groupContext` decorator allows the developer to wrap groups of certain
 * types of elements in order to provide them with the same name without having
 * to manually wire each element. Supported elements include `Checkbox` and
 * `Radio`, as well as, generally, their subclasses.
 *
 * The `groupContext` decorator does not provide values to its children. Rather,
 * that remains the responsibility of the wrapping `Form`.
 *
 * @returns {Function} - A new React Element wrapping the given component
 */
export default () => (Wrapped) => {
    class AsGrouped extends React.PureComponent {
        static Wrapped = Wrapped;

        static propTypes = {
            forwardedRef: PropTypes.any
        };

        render() {
            const {
                forwardedRef,
                ...props
            } = this.props;
            return (
                <FieldGroupContext.Consumer>
                    { (ctx) => (
                        <Wrapped name={ ctx } ref={ forwardedRef } { ...props } />
                    ) }
                </FieldGroupContext.Consumer>
            );
        }
    }
    const Forwarded = React.forwardRef((props, ref) => (
        <AsGrouped { ...props } forwardedRef={ ref } />
    ));
    Forwarded.Wrapped = AsGrouped;
    return Forwarded;
};
