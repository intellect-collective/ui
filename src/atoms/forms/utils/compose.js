// Lifted from the big boys at
// https://github.com/reduxjs/redux/blob/b4fb08133c95094a4b293a9ab434d1d5dd657527/src/compose.js
export default (...funcs) => {
    if (funcs.length === 0) {
        return (arg) => (arg);
    }

    if (funcs.length === 1) {
        return funcs[0];
    }

    return funcs.reduce((a, b) => (...args) => (a(b(...args))));
};
