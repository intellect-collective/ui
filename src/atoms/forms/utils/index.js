export { default as asCheckable } from './asCheckable';
export { default as asField } from './asField';
export { default as asGroupable } from './asGroupable';
export { default as compose } from './compose';
export { default as normalizeItems } from './normalizeItems';
