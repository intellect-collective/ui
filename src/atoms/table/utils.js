export const getSort = (sort) => {
    if (!sort || sort === '') {
        return [undefined, undefined];
    }

    if (sort.charAt(0) === '-') {
        return [sort.substr(1), '-'];
    }

    if (sort.charAt(0) === '+') {
        return [sort.substr(1), '+'];
    }

    return [sort, '+'];
};

export const ifSort = (col, sort, ascending, descending, unsorted) => {
    const [sortcol, dir] = getSort(sort);
    if (sortcol !== col) {
        return unsorted;
    }
    if (dir === '+') {
        return ascending;
    }
    if (dir === '-') {
        return descending;
    }
    return unsorted;
};
