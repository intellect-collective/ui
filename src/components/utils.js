import clsx from 'clsx';
import pick from 'lodash/pick';

/**
 * Builds a component's className attribute from the information given.
 */
export function classify({
	className,
	classes,
	color
}, ancillary) {
	return clsx(className, classes, {
		[`color-${ color }`]: !!color
	}, ancillary);
};


export function slot(children, fn) {
	if (!Array.isArray(children)) {
		return;
	}

	return children.filter(fn);
}


export function html_props(props) {
	return pick(props, ['style']);
}