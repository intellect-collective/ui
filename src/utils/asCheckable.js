import React from 'react';
import PropTypes from 'prop-types';

const noop = () => {};

const asCheckable = ({
    /**
     * The prop that the `Field` component uses to pass the form value down.
     * Checkboxes, radios, etc where the value attribute does not equal the
     * visible state often require the form value to be passed down through a
     * different prop for comparison purposes.
     */
    formValueProp = 'formValue',
    /**
     * Whether or not to allow multiple selection. Checkboxes allow for multiple
     * selection, but radios do not. The `Checkable` component will determine
     * the checked state the same way, but will only allow the handler to
     * persist a boolean or single string value.
     */
    multiple = true
} = {}) => (Wrapped) => {
    class Checkable extends React.PureComponent {
        static Wrapped = Wrapped;

        static propTypes = {
            name: PropTypes.string.isRequired,
            value: PropTypes.oneOfType([
                PropTypes.string,
                PropTypes.number
            ]),
            checked: PropTypes.bool,
            boolean: PropTypes.bool,
            onChange: PropTypes.func,
            onBlur: PropTypes.func,
            forwardedRef: PropTypes.any
        };

        static defaultProps = {
            onChange: noop,
            onBlur: noop
        }

        constructor() {
            super();
            this.onChange = this.onChange.bind(this);
        }

        onChange(ev) {
            console.log('here')
            const {
                boolean,
                value,
                onChange,
                [formValueProp]: formValue
            } = this.props;

            let result;
            if (typeof value === 'undefined' || boolean) {
                result = !this.checked;
            } else if (!multiple) {
                result = this.checked ? null : value;
            } else if (this.checked) {
                result = [].concat(formValue).filter((o) => (o !== value && typeof o !== 'undefined'));
            } else {
                result = [value].concat(formValue).filter((o) => (typeof o !== 'undefined'));
            }

            onChange(ev, result);
        }

        get checked() {
            const {
                boolean,
                checked,
                value,
                [formValueProp]: formValue
            } = this.props;
            if (typeof checked !== 'undefined') {
                return !!checked;
            }
            if (typeof value === 'undefined' || boolean) {
                return typeof checked !== 'undefined' ? checked : (formValue === 'true' || formValue === true);
            }
            if (Array.isArray(formValue)) {
                return formValue.indexOf(value) !== -1;
            }
            return formValue === value;
        }

        render() {
            const {
                boolean,
                forwardedRef,
                [formValueProp]: formValue,
                ...props
            } = this.props;
            return (
                <Wrapped { ...props }
                        checked={ this.checked }
                        onChange={ this.onChange }
                        ref={ forwardedRef } />
            );
        }
    }

    const Forwarded = React.forwardRef((props, ref) => (
        <Checkable { ...props } forwardedRef={ ref } />
    ));
    Forwarded.Wrapped = Checkable;
    return Forwarded;
};

export default asCheckable;
