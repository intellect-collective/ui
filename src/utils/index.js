export { default as asCheckable } from './asCheckable';
export { default as asField } from './asField';
export { default as asGroupable } from './asGroupable';
export { default as compose } from './compose';
export { default as get } from './get';
export { default as getSort } from './getSort';
export { default as getWrapped } from './getWrapped';
export { default as ifSort } from './ifSort';
export { default as is } from './is';
export { default as memoize } from './memoize';
export { default as memoizeLast } from './memoizeLast';
export { default as normalizeItems } from './normalizeItems';
export { default as pushUnique } from './pushUnique';
export { default as shallowEqual } from './shallowEqual';
export { default as withContext } from './withContext';
