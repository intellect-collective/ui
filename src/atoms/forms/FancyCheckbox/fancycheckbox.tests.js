import React from 'react';
import { FancyCheckbox } from '../../../..';
import { checkableTests } from '../../../../test/utils';

describe('FancyCheckboxComponent', () => {
    checkableTests(FancyCheckbox);
});
