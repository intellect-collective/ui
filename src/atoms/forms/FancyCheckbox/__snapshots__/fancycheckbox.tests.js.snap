// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FancyCheckbox Basic changeable tests Should accept valid props 1`] = `
<GroupContext
  name="test"
  onChange={[Function]}
  value="fire"
>
  <FormContext
    name="test"
    onChange={[Function]}
    value="fire"
  >
    <FormContextInner
      ctx={null}
      name="test"
      onChange={[Function]}
      value="fire"
    >
      <Checkable
        name="test"
        onChange={[Function]}
        value="fire"
      >
        <Changeable
          checked={false}
          name="test"
          onChange={[Function]}
          value="fire"
        >
          <FancyCheckbox
            checked={false}
            name="test"
            onChange={[Function]}
            value="fire"
          >
            <div
              className="fancycheck"
            >
              <input
                checked={false}
                name="test"
                onChange={[Function]}
                type="checkbox"
                value="fire"
              />
              <ins
                aria-checked={false}
                onClick={[Function]}
                onKeyDown={[Function]}
                role="checkbox"
                tabIndex={-1}
              />
            </div>
          </FancyCheckbox>
        </Changeable>
      </Checkable>
    </FormContextInner>
  </FormContext>
</GroupContext>
`;
